### PR TITLE
Internal: JS API - Partial refactor [ED-2084] 

### PR DESCRIPTION
**Pull request by iNewLegend**
- JS API command base shouldn't know about the editor, separate the editor commands
- The running of commands and handling of the response should be from the commands manager, not from the command itself
- Added a global trace that combined all commands managers traces (for easy debugging)
- Each command class should know its own component and full command name on a static level.

Next:
- Todo: Route Hooks
- Todo: Async Hooks
- Todo: Maybe call `initialize` after `validateArgs`.
- Todo: Update Docs.

### CODE REVIEW

**Review by matipojo — CHANGES_REQUESTED** [review]
Please resolve the conflicts

File: assets/dev/js/editor/editor-document.js [hunk 1]
```diff
@@ -1,14 +1,38 @@
...
 
-	// TODO: Remove, BC Since 2.9.0.
-	elementor.saver = $e.components.get( 'document/save' );
-} );
+// TODO: Remove, BC.
+$e.modules.document = {
+	get CommandHistory() {
+		elementorCommon.helpers.softDeprecated(
+			'$e.modules.document.CommandHistory',
+			'3.5.0',
```
Comment (matipojo) [inline assets/dev/js/editor/editor-document.js:21]:
```suggestion
			'3.7.0',
```

File: assets/dev/js/editor/editor-document.js [hunk 2]
```diff
@@ -1,14 +1,38 @@
...
+			'$e.modules.editor.document.CommandHistoryBase'
+		);
+
+		return $e.modules.editor.document.CommandHistoryBase;
+	},
+
+	get CommandHistoryDebounce() {
+		elementorCommon.helpers.softDeprecated(
+			'$e.modules.CommandHistoryDebounce',
+			'3.5.0',
```
Comment (matipojo) [inline assets/dev/js/editor/editor-document.js:31]:
```suggestion
			'3.7.0',
```

File: modules/web-cli/assets/js/core/commands.js [hunk 3]
```diff
@@ -1,31 +1,31 @@
...
+	constructor( ... args ) {
+		super( ... args );
 
 		this.current = {};
 		this.currentArgs = {};
 		this.currentTrace = [];
 		this.commands = {};
 		this.components = {};
-
-		this.classes = {};
```
Comment (matipojo) [inline modules/web-cli/assets/js/core/commands.js:20]: BC

File: modules/web-cli/assets/js/core/commands.js [hunk 4]
```diff
@@ -243,18 +243,39 @@ export default class Commands extends CommandsBackwardsCompatibility {
...
+	 * Responsible to add current command to trace and trigger 'run:before' event.
+	 * Run before command.
 	 *
-	 * @returns {boolean} dependency result
+	 * @param {string} command
+	 * @param {{}} args
 	 */
 	beforeRun( command, args = {} ) {
+		const component = this.getComponent( command ),
+			container = component.getRootContainer();
```
Comment (matipojo) [inline modules/web-cli/assets/js/core/commands.js:254]:
```suggestion
			serviceName = component.getServiceName();
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/web-cli/assets/js/core/commands.js (hunk 4 above)
Comment (iNewLegend) [reply modules/web-cli/assets/js/core/commands.js:254]: https://github.com/elementor/elementor/pull/18163

File: modules/web-cli/assets/js/core/commands.js (hunk 3 above)
Comment (iNewLegend) [reply modules/web-cli/assets/js/core/commands.js:20]: https://github.com/elementor/elementor/pull/18163

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/web-cli/assets/js/core/commands.js [hunk 5]
```diff
@@ -1,31 +1,48 @@
...
+					'3.7.0',
+					'$e.commands.getCommandClass(), $e.commandsInternal.getCommandClass(), $e.data.getCommandClass(), $e.routes.getCommandClass() according to the requested command infra-structure,'
+				);
+
+				return {
+					... $e.commands.commands,
+					... $e.commandsInternal.commands,
+					... $e.data.commands,
+					... $e.routes.commands,
+				};
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply modules/web-cli/assets/js/core/commands.js:36]: https://github.com/elementor/elementor/blob/master/core/common/assets/js/api/modules/component-base.js#L351

File: tests/qunit/tests/assets/dev/js/editor/document/command-bases/command-history-base.spec.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+						const instance = new CommandHistoryBaseMock( {} );
+
+						instance.getHistory( {} );
+					},
+					new Error( 'CommandHistoryBaseMock.getHistory() should be implemented, please provide \'getHistory\' functionality.' )
+				);
+			} );
+
+			QUnit.test( 'onCatchApply()`', ( assert ) => {
+				const fakeHistory = class extends CommandHistoryBaseMock {
```
Comment (matipojo) [inline tests/qunit/tests/assets/dev/js/editor/document/command-bases/command-history-base.spec.js:22]: AAA

File: tests/qunit/tests/assets/dev/js/editor/document/component.spec.js [hunk 7]
```diff
@@ -59,11 +59,11 @@ QUnit.module( 'Component: document', () => {
 			assert.equal( elementor.documents.getCurrent().history.getItems().length, 3,
 				'Master document still have "3" items in history.' );
 
-			// Put back saved document, to current.
-			elementor.documents.setCurrent( documentBeforeTest );
-			eData.emptyFetch();
-
 			done();
+
```
Comment (matipojo) [inline tests/qunit/tests/assets/dev/js/editor/document/component.spec.js:63]: Explain why the `done` is before the real `done`

File: tests/qunit/tests/assets/dev/js/editor/document/elements/component.spec.js [hunk 8]
```diff
@@ -1,6 +1,7 @@
 import DocumentHelper from '../helper';
-import ElementsHelper from './helper';
 import HistoryHelper from '../history/helper';
+import ElementsHelper from './helper';
```
Comment (matipojo) [inline tests/qunit/tests/assets/dev/js/editor/document/elements/component.spec.js:3]: Restore

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/web-cli/assets/js/core/routes.js [hunk 9]
```diff
@@ -65,34 +64,56 @@ export default class Routes extends Commands {
...
 
 		if ( this.is( route, args ) && ! args.refresh ) {
 			return false;
 		}
 
+		return this.getComponent( route )?.isOpen;
+	}
+
+	run( route, args = {} ) {
+		if ( ! this.validateRun( route, args ) ) {
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply modules/web-cli/assets/js/core/routes.js:80]: Handled: https://github.com/elementor/elementor/pull/18221

File: modules/web-cli/assets/js/core/routes.js [hunk 10]
```diff
@@ -65,34 +64,56 @@ export default class Routes extends Commands {
...
+			const component = this.getComponent( route );
+
+			if ( ! component.isOpen || args.reOpen ) {
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply modules/web-cli/assets/js/core/routes.js:83]: Handled: https://github.com/elementor/elementor/pull/18221

File: modules/web-cli/assets/js/core/routes.js [hunk 11]
```diff
@@ -65,34 +64,56 @@ export default class Routes extends Commands {
...
+			container = component.getServiceName(),
 			oldRoute = this.current[ container ];
 
 		if ( oldRoute ) {
 			this.getComponent( oldRoute ).onCloseRoute( oldRoute );
 		}
 
-		if ( ! component.isOpen || args.reOpen ) {
-			component.isOpen = component.open( args );
+		if ( args.onBefore ) {
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply modules/web-cli/assets/js/core/routes.js:104]: Handled: https://github.com/elementor/elementor/pull/18221

File: modules/web-cli/assets/js/core/routes.js [hunk 12]
```diff
@@ -125,16 +146,24 @@ export default class Routes extends Commands {
...
 
 	// Don't clear current route.
-	afterRun( route, args ) {
-		this.getComponent( route ).onRoute( route, args );
+	afterRun( route, args, results = undefined ) {
+		const component = this.getComponent( route );
+
+		component.onRoute( route, args );
+
+		if ( args.onAfter ) {
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply modules/web-cli/assets/js/core/routes.js:154]: Handled: https://github.com/elementor/elementor/pull/18221

File: modules/web-cli/assets/js/modules/command-base.js [hunk 13]
```diff
@@ -62,6 +50,8 @@ export default class CommandBase extends ArgsObject {
 	 * @throws {Error}
 	 */
 	requireContainer( args = this.args ) {
+		elementorCommon.helpers.softDeprecated( 'requireContainer', '3.6.0', 'Extend `$e.modules.editor.CommandContainerBase` or `$e.modules.editor.CommandContainerInternalBase`' );
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply modules/web-cli/assets/js/modules/command-base.js:53]: Handled: https://github.com/elementor/elementor/pull/18221

File: modules/web-cli/assets/js/modules/command-callback.js [hunk 14]
```diff
@@ -0,0 +1,25 @@
+import CommandBase from './command-base';
+
+/**
+ * To support pure callbacks in the API(commands.js), to ensure they have registered with the proper context.
+ */
+export default class CommandCallback extends CommandBase {
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply modules/web-cli/assets/js/modules/command-callback.js:6]: Handled: https://github.com/elementor/elementor/pull/18221

File: modules/web-cli/assets/js/modules/component-base.js [hunk 15]
```diff
@@ -149,8 +160,33 @@ export default class ComponentBase extends Module {
...
 	}
 
-	registerCommand( command, callback ) {
-		$e.commands.register( this, command, callback );
+	/**
+	 * @param {string} command
+	 * @param {(function()|typeof CommandInfra)} context
+	 * @param commandsAPI
+	 */
+	registerCommand( command, context, commandsAPI = $e.commands ) {
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply modules/web-cli/assets/js/modules/component-base.js:168]: Handled: https://github.com/elementor/elementor/pull/18221

File: tests/qunit/tests/assets/dev/js/editor/document/command-bases/command-history-base.spec.js (hunk 6 above)
Comment (iNewLegend) [reply tests/qunit/tests/assets/dev/js/editor/document/command-bases/command-history-base.spec.js:22]: Handled: https://github.com/elementor/elementor/pull/18221

File: tests/qunit/tests/assets/dev/js/editor/document/component.spec.js (hunk 7 above)
Comment (iNewLegend) [reply tests/qunit/tests/assets/dev/js/editor/document/component.spec.js:63]: Handled: https://github.com/elementor/elementor/pull/18221

File: tests/qunit/tests/assets/dev/js/editor/document/elements/component.spec.js (hunk 8 above)
Comment (iNewLegend) [reply tests/qunit/tests/assets/dev/js/editor/document/elements/component.spec.js:3]: Handled: https://github.com/elementor/elementor/pull/18221

**Pull request merged**